build/pack.c

```diff
@@ -726,16 +727,37 @@ static rpmRC packageBinary(rpmSpec spec, Package pkg, const char *cookie, int ch
     return rc;
 }
 
+static int compareBinaries(const void *p1, const void *p2) {
+    Package pkg1 = *(Package *)p1;
+    Package pkg2 = *(Package *)p2;
+    uint64_t size1 = headerGetNumber(pkg1->header, RPMTAG_SIZE);
+    uint64_t size2 = headerGetNumber(pkg2->header, RPMTAG_SIZE);
+    return size1 - size2;
+}
+
+/*
+ * Run binary creation in parallel, with task priority based on package size
+ * (largest first) to help achieve an optimal load distribution.
+ */
 rpmRC packageBinaries(rpmSpec spec, const char *cookie, int cheating)
 {
     rpmRC rc = RPMRC_OK;
     Package pkg;
+    Package *tasks;
+    int i = 0;
+
+    for (pkg = spec->packages; pkg != NULL; pkg = pkg->next)
+        i++;
+    tasks = xcalloc(i, sizeof(Package));
+    for (i = 0, pkg = spec->packages; pkg != NULL; i++, pkg = pkg->next)
+        tasks[i] = pkg;
+    qsort(tasks, i, sizeof(Package), compareBinaries);
```

Comment:
Do use separate variables for indexing and things like total package count, (re)using an 'i' for multiple different purposes only makes the code harder to follow for now good reason. Using loop-local index variables is good way to ensure there are no side-effects from previous rounds, and also makes it obvious to the reader as well.

```
    int npkgs = 0;

    for (pkg = spec->packages; pkg != NULL; pkg = pkg->next)
        npkgs++;
    tasks = xcalloc(npkgs, sizeof(Package));
    for (int i = 0, pkg = spec->packages; pkg != NULL; i++, pkg = pkg->next)
        tasks[i] = pkg;
    qsort(tasks, npkgs, sizeof(Package), compareBinaries);
```